src/bob/resource/internals.clj

```diff
@@ -0,0 +1,85 @@
+;   This file is part of Bob.
+;
+;   Bob is free software: you can redistribute it and/or modify
+;   it under the terms of the GNU General Public License as published by
+;   the Free Software Foundation, either version 3 of the License, or
+;   (at your option) any later version.
+;
+;   Bob is distributed in the hope that it will be useful,
+;   but WITHOUT ANY WARRANTY; without even the implied warranty of
+;   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+;   GNU General Public License for more details.
+;
+;   You should have received a copy of the GNU General Public License
+;   along with Bob. If not, see <http://www.gnu.org/licenses/>.
+
+(ns bob.resource.internals
+  (:require [aleph.http :as http]
+            [failjure.core :as f]
+            [bob.execution.internals :refer [pull build copy-dir!]]
+            [bob.util :refer [unsafe!]])
+  (:import (java.util.zip ZipInputStream)
+           (java.io File)
+           (java.nio.file Files)
+           (java.nio.file.attribute FileAttribute)))
+
+(defn- rm-r!
+  "Recursively deletes a directory."
+  [dir & [silently]]
+  (unsafe!
+    (letfn [(delete-f [^File file]
+              (when (.isDirectory file)
+                (doseq [child-file (.listFiles file)]
+                  (delete-f child-file)))
+              (clojure.java.io/delete-file file silently))]
+      (delete-f (clojure.java.io/file dir)))))
+
+(defn- extract-zip!
+  "Takes a java.util.zip.ZipInputStream `zip-stream` and extracts the content to the `out-dir`."
+  [^ZipInputStream zip-stream out-dir]
+  (unsafe!
+    (with-open [stream zip-stream]
+      (loop [entry (.getNextEntry stream)]
+        (when entry
+          (let [savePath (str out-dir File/separatorChar (.getName entry))
+                saveFile (File. savePath)]
+            (if (.isDirectory entry)
+              (if-not (.exists saveFile)
+                (.mkdirs saveFile))
+              (let [parentDir (File.
+                                (.substring
```

Comment:
this part of code might deserve separate function, like `dirname` in python: https://docs.python.org/3/library/os.path.html#os.path.dirname